### PR TITLE
removed Lock() from ThreadManager.py

### PR DESCRIPTION
**Pull request by RickyDanipog**
added new class ThreadManagerGUI, used for the GUI implementation.
Model.learning_model has been changed,
for lower coupling :)

### CODE REVIEW

File: Algorithm/ThreadManager.py [hunk 1]
```diff
@@ -68,13 +59,37 @@ def return_best_score(cls) -> Tuple[int, dict]:
         """
         :return: Tuple of best result and which set of parameters
         """
-        return sorted(cls.results, key=lambda item: item.score, reverse=False)[0]
+        return sorted(cls.results.values(), key=lambda res: res.score, reverse=False)[0]
+
+
+class ThreadManagerGUI(ThreadManager):
```
Comment (royjan) [inline Algorithm/ThreadManager.py:65]: Does it really necessary to have 2 diff classes ThreadManager?

File: main.py [hunk 2]
```diff
@@ -14,9 +14,11 @@
 from matplotlib import pyplot as plt
 from sklearn.tree import DecisionTreeRegressor
 from sklearn.svm import SVC
+from sklearn.linear_model import LinearRegression
```
Comment (royjan) [inline main.py:17]: We won't send this file next time.. should we update it?

File: Algorithm/ThreadManager.py [hunk 3]
```diff
@@ -5,36 +5,29 @@
...
+from threading import Thread
+from typing import Tuple
 
 from Algorithm.Model import learning_model
 from Utils.Log import writer
 
 TIMEOUT_THREAD = 10  # seconds
 
 
-@contextmanager
```
Comment (royjan) [inline Algorithm/ThreadManager.py:19]: How do we handle with dead locks?

File: Algorithm/ThreadManager.py [hunk 4]
```diff
@@ -68,13 +59,37 @@ def return_best_score(cls) -> Tuple[int, dict]:
...
+
+    sorted_results = []
+
+    @classmethod
+    def wait_for_all_threads(cls, **kwargs):
+        """
+        A manual thread.join(), needed for components that will need to be triggered when a thread ends his task
+        @param kwargs: components needed to react to the thread's training process
+        """
+        check_checkbox = kwargs['check_checkbox']
```
Comment (royjan) [inline Algorithm/ThreadManager.py:75]: check_checkbox = kwargs.get('check_checkbox')
if check_checkbox:

more generic I think..

**Review by royjan — CHANGES_REQUESTED** [review]

File: Algorithm/ThreadManager.py (hunk 3 above)
Comment (RickyDanipog) [reply Algorithm/ThreadManager.py:19]: there are no deadlocks, each train is independent on others.

File: main.py (hunk 2 above)
Comment (RickyDanipog) [reply main.py:17]: keep it for debugging and testing until the final review

File: Algorithm/ThreadManager.py (hunk 1 above)
Comment (RickyDanipog) [reply Algorithm/ThreadManager.py:65]: No ,but I divided the class so that we can use it on GUI less runs

File: Algorithm/ThreadManager.py (hunk 4 above)
Comment (RickyDanipog) [reply Algorithm/ThreadManager.py:75]: ok.

**Review by royjan — APPROVED** [review]
Good job Ricky! 👑

**Pull request merged**